src/bos/common/clients/endpoints/base_generic_endpoint.py

```diff
@@ -0,0 +1,119 @@
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+from abc import ABC, abstractmethod
+import logging
+from typing import Generic, TypeVar
+
+import requests
+
+from .defs import RequestData, RequestsMethod
+from .exceptions import ApiResponseError
+from .request_error_handler import BaseRequestErrorHandler, RequestErrorHandler
+
+LOGGER = logging.getLogger(__name__)
+
+RequestReturnT = TypeVar('RequestReturnT')
+
+
+class BaseGenericEndpoint(ABC, Generic[RequestReturnT]):
+    """
+    This base class provides generic access to an API endpoint.
+    RequestReturnT represents the type of data this API will return.
+    Most often this will be the Json data from the response body, but in some
+    cases (like with BSS), we are after something else.
+
+    Exceptions are handled by a separate class, since different API clients
+    may want to handle these differently.
+    """
+    BASE_ENDPOINT: str = ''
+    ENDPOINT: str = ''
+    error_handler: BaseRequestErrorHandler = RequestErrorHandler
+
+    def __init__(self, session: requests.Session):
+        super().__init__()
+        self.session = session
+
+    @classmethod
+    @abstractmethod
+    def format_response(cls, response: requests.Response) -> RequestReturnT:
+        ...
+
+    @classmethod
+    def base_url(cls) -> str:
+        return f"{cls.BASE_ENDPOINT}/{cls.ENDPOINT}"
+
+    @classmethod
+    def url(cls, uri: str) -> str:
+        base_url = cls.base_url()
+        if not uri:
+            return base_url
+        if uri[0] == '/' or base_url[-1] == '/':
+            return f"{base_url}{uri}"
+        return f"{base_url}/{uri}"
+
+    def request(self,
+                method: RequestsMethod,
+                /,
+                *,
+                uri: str = "",
+                **kwargs) -> RequestReturnT:
+        url = self.url(uri)
+        LOGGER.debug("%s %s (kwargs=%s)", method.__name__.upper(), url, kwargs)
```

Comment:
use string formatting f""

Reply:
You're not supposed to do that with logging statements. That's because the logger will only evaluate the string if that level of logging is enabled (whereas if you use a f-string, the logger will always format it, even if it ends up not being logged). pylint will complain any time it seems an f-string used with a logging function